controllers/azuremachine_controller.go

```diff
@@ -51,9 +51,33 @@ import (
 // AzureMachineReconciler reconciles a AzureMachine object
 type AzureMachineReconciler struct {
 	client.Client
-	Log              logr.Logger
-	Recorder         record.EventRecorder
-	ReconcileTimeout time.Duration
+	Log                        logr.Logger
+	Recorder                   record.EventRecorder
+	ReconcileTimeout           time.Duration
+	azureMachineServiceFactory func(machineScope *scope.MachineScope, clusterScope *scope.ClusterScope) *azureMachineService
```

Comment:
Should `azureMachineServiceFactory` be exported? If you are going to provide variadic options for injecting the service, seems like a lot of work to not expose it on the public API? If this is only to be used in testing, perhaps, it would be best to only set `azureMachineServiceFactory` via setting the field in the struct directly and forgo all the work to add variadic options.

Reply:
>I'm not clear that the azureMachineServiceFactory will be set anywhere besides tests. If the ability to inject an azureMachineServiceFactory for testing purposes, perhaps, that can be done simply by having a test within the same package as this struct and set the struct field directly.

Addressing this as well since it's on a similar note, I didn't want to export `azureMachineServiceFactory` for a few reasons :
1. It's also used in `main.go, and so we'll then need to export `newAzureMachineService` as well, and I thought it might make main.go cumbersome.
2. Better to have an api to initialize rather than setting the struct values directly. Provides consistency and easier to test. 
3. Easier to set defaults.

I agree that the variadic options are not needed right now, but just added it to make sure that there is a way to override defaults in case something changes in future. wdyt?

Reply:
This is the only optional we've needed to add. I'm hesitant to add variadic options for this only because we might add others in the future. Since it's not on the public API for `NewAzureMachineReconciler` func, I'd vote to keep it simple. In the `NewAzureMachineReconciler`, we can set the default as you are doing, and in tests, just replace the default value set in the constructor. Everything would work the same as today, but also provide us the ability to replace the default in tests.